### ISSUE TITLE
Buffer overflows (found with AFL)

### ISSUE THREAD

**Issue by MichielP1807** [lecram/gifdec#15]
Thanks for making this great library! It was fun learning a bit about how gifs work!

We fuzzed this library for a school project and we found quite a few buffer overflow vulnerabilities and some integer overflow vulnerabilities. These can cause a variety of different errors, from segmentation fault to all kinds of free and malloc based memory errors. There also were quite a few potentially uninitialized values. We made an attempt at fixing the buffer overflows and integer overflows, but we were not sure how to fix all the potentially uninitialized values that can still crash this library with certain input gifs.

- One common issue that caused buffer overflows, is that frames can go out of bounds within the gif. If the `gif->fx`, `gif->fy`, `gif->fw`, or `gif->fh` are too large, then we will be reading and writing outside of the `gif` buffer. To fix this, we have changed some of the for loops to make sure that the x and y coordinates do not go outside of the width and height of the gif.
- Another issue is an integer overflow when the width and height of a gif are very large. The `calloc(1, sizeof(*gif) + 4 * width * height)` that allocates memory for the `gif` buffer will sometimes allocated not enough memory because the value of `sizeof(*gif) + 4 * width * height` can overflow. In our attempt to fix this, we have separated this `calloc` into 3 separate `calloc`'s: one for the `gif` struct, one for the `gif->canvas`, and one for the `gif->frame`.
- In the `add_entry` function, there can be a buffer overflow when `table->nentries` is not set properly. The value of `table->nentries` is computed either `(1 << key_size) + 2` or `(1 << (key_size - 1)) + 2`. We added some extra checks to make sure that the `key_size` is not too large, since the computations overflow when `key_size` is larger or equal to 31 or 32 respectively. We are not entirely sure that this is indeed a good assumption that `key_size` should be smaller than 31 or 32 at these places, so if you have any comments, please let us know!

**Comment by lecram** [comment]
> We fuzzed this library for a school project

That's awesome!

> * One common issue that caused buffer overflows, is that frames can go out of bounds within the gif. If the `gif->fx`, `gif->fy`, `gif->fw`, or `gif->fh` are too large, then we will be reading and writing outside of the `gif` buffer. To fix this, we have changed some of the for loops to make sure that the x and y coordinates do not go outside of the width and height of the gif.

Of the 3 issues reported, this is the hardest to fix without performance degradation. I'd like to avoid doing bounds-checking inside the inner loop in `read_image_data()` because it runs once per pixel and is already too slow on some hardware. My suggestion would be checking `fx`, `fy`, `fw` and `fh` as soon as they are read, in `read_image()`, then changing `fw` and `fh` if necessary to avoid overflow. This would happen only once per frame. The function `read_image_data()` is already capable of discarding extra pixels from image data so that should be OK.

> * Another issue is an integer overflow when the width and height of a gif are very large. The `calloc(1, sizeof(*gif) + 4 * width * height)` that allocates memory for the `gif` buffer will sometimes allocated not enough memory because the value of `sizeof(*gif) + 4 * width * height` can overflow. In our attempt to fix this, we have separated this `calloc` into 3 separate `calloc`'s: one for the `gif` struct, one for the `gif->canvas`, and one for the `gif->frame`.

The proposed solution for this one looks good to me. For extra points I'd keep `canvas` and `frame` in the same block, swapping their position:
```
    gif->frame = calloc(4, width * height);
    gif->canvas = &gif->frame[width * height];
```
I generally try to put multiple buffers in the same block to ensure they're contiguous in memory, optimizing cache performance. It's also easier to cleanup.

> * In the `add_entry` function, there can be a buffer overflow when `table->nentries` is not set properly. The value of `table->nentries` is computed either `(1 << key_size) + 2` or `(1 << (key_size - 1)) + 2`. We added some extra checks to make sure that the `key_size` is not too large, since the computations overflow when `key_size` is larger or equal to 31 or 32 respectively. We are not entirely sure that this is indeed a good assumption that `key_size` should be smaller than 31 or 32 at these places, so if you have any comments, please let us know!

`key_size` is read as an unsigned byte in the beginning of `read_image_data()`. The expected value is in the range 2..8 inclusive, but of course for an ill-formed input it could be anything in the range 0..255. In `read_image_data()`, `key_size` is incremented by one every time `nentries` crosses a power of two (e.g. when `nentries` reaches 1024, `key_size` is incremented from 9 to 10). However, when `nentries` reaches 2^12 (= 0x1000 = 4096), the table is full and we stop calling `add_entry()`. `key_size` can also be reset to its initial value when a "clear code" is found. So if the input *is* in the range 2..8, `key_size` will never go outside the range 2..12. Therefore to avoid the overflow it's sufficient to check `key_size`'s initial value as soon as it's read in the beginning of `read_image_data()`, which should be in the range 2..8. This also happens only once per frame.


About the potentially uninitialized values, please feel free to open a separate issue with more details if you're interested in discussing that.

**Comment by MichielP1807** [comment]
Thanks for your feedback! In the new PR I've re-implemented the fixes, and made sure to turn off auto-formatting this time.

> Of the 3 issues reported, this is the hardest to fix without performance degradation. I'd like to avoid doing bounds-checking inside the inner loop in `read_image_data()` because it runs once per pixel and is already too slow on some hardware. My suggestion would be checking `fx`, `fy`, `fw` and `fh` as soon as they are read, in `read_image()`, then changing `fw` and `fh` if necessary to avoid overflow.

Indeed, this one has a larger impact on performance compared to the other fixes since we're currently checking every pixel to be in bounds which is not ideal. Your suggestion seems like it would fix the other `gif->frame` buffer overflows as well, which is nice. If we do the following:

```c
    gif->fw = MIN(gif->fw, gif->width - gif->fx);
    gif->fh = MIN(gif->fh, gif->height - gif->fy);
```
Then the dimensions of the frame should never go outside of the dimensions of the gif. That is, unless `gif->fx` or `gif->fy` are absurdly big, in which case we could get a problematic integer underflow. If that's the case I think it would be best to just return a failure. So I think we would also need to add something like:
```c
    if (gif->fx >= gif->width || gif->fy >= gif->height)
        return -1;
```
> I generally try to put multiple buffers in the same block to ensure they're contiguous in memory, optimizing cache performance. It's also easier to cleanup.

This is an interesting point we hadn't considered! I've updated the PR with the changes as you suggested.

> `key_size` is read as an unsigned byte in the beginning of `read_image_data()`. The expected value is in the range 2..8 inclusive, but of course for an ill-formed input it could be anything in the range 0..255. In `read_image_data()`, `key_size` is incremented by one every time `nentries` crosses a power of two (e.g. when `nentries` reaches 1024, `key_size` is incremented from 9 to 10). However, when `nentries` reaches 2^12 (= 0x1000 = 4096), the table is full and we stop calling `add_entry()`. `key_size` can also be reset to its initial value when a "clear code" is found. So if the input is in the range 2..8, `key_size` will never go outside the range 2..12. Therefore to avoid the overflow it's sufficient to check `key_size`'s initial value as soon as it's read in the beginning of `read_image_data()`, which should be in the range 2..8. 

Thanks for the explanation! Checking that `key_size` is in the range 2..8 at the beginning of `read_image_data()` indeed seems to have fixed the issue.

I have ran AFL again for a bit with these improved fixes, and it seems that the only crashes that we are still getting are related to uninitialized values, just like we did as with our previous less efficient fixes. The crashes that can still happen are ones for which Valgrind warns us about both the use of uninitialized values and conditional jumps or moves depending on uninitialized values. I'll open a separate issue for this.

**Comment by lecram** [comment]
Thank you for the update. It's merged.
